File: src/net/merayen/elastic/backend/architectures/local/nodes/midi_1/LProcessor.java
package net.merayen.elastic.backend.architectures.local.nodes.midi_1;

import java.util.List;

import net.merayen.elastic.backend.architectures.local.LocalProcessor;
import net.merayen.elastic.backend.architectures.local.lets.MidiOutlet;
import net.merayen.elastic.util.Postmaster.Message;

public class LProcessor extends LocalProcessor {
	@Override
	protected void onInit() {}

	@Override
	protected void onPrepare() {}

	@Override
	protected void onProcess() {
		MidiOutlet outlet = (MidiOutlet)getOutlet("out");
		if(outlet != null) {
			List<short[]> midi_from_ui = ((LNode)getLocalNode()).midi_from_ui;
			outlet.midi[0] = midi_from_ui.toArray(new short[midi_from_ui.size()][]);
			outlet.written = buffer_size;
			outlet.push();
		}
	}

	@Override
	protected void onMessage(Message message) {}

	@Override
	protected void onDestroy() {}
}

File: src/main/kotlin/net/merayen/elastic/backend/architectures/local/nodes/midi_1/LProcessor.kt
package net.merayen.elastic.backend.architectures.local.nodes.midi_1

import net.merayen.elastic.backend.architectures.local.GroupLNode
import net.merayen.elastic.backend.architectures.local.LocalProcessor
import net.merayen.elastic.backend.architectures.local.lets.MidiOutlet

class LProcessor : LocalProcessor() {
	private var playStartedCount: Long = 0
	private var midiDataRevision = 0L

	override fun onInit() {}

	override fun onPrepare() {}

	override fun onProcess() {
		val outlet = getOutlet("out") as MidiOutlet?
		if (outlet != null) {
			val parent = localNode.parent as GroupLNode

			val startBeat = parent.getCursorPosition()

			val newPlayStartedCount = parent.playStartedCount()
			if (newPlayStartedCount != playStartedCount)
				playStartedCount = newPlayStartedCount

			val stopBeat = if (parent.isPlaying()) {
				startBeat + (parent.getCurrentFrameBPM() * (buffer_size / sampleRate.toDouble())) / 60.0
			} else {
				startBeat
			}

			if (parent.isPlaying())
				; // TODO println("Playing! startBeat=${startBeat}, stopBeat=${stopBeat}")

			val lnode = (localNode as LNode)
			val inputMidi = lnode.inputMidi
			val midiData = lnode.midiData

			if (inputMidi != null) // TODO write any note from MidiData too
				outlet.addMidi(0, inputMidi) // TODO shouldn't quantize to beginning of frame

			if (midiData != null) {
				if (midiData.revision != midiDataRevision) {
					println("Midi data received!")
					midiDataRevision = midiData.revision
				}
			}

			outlet.push()
		}
	}

	override fun onDestroy() {}
}
